Integrate the MicroProfile profiling library

yuriks opened the pull request:
This brings goodies such as a configurable user interface and multi-threaded timeline view.

Screenshot:

![Detailed View](http://i.imgur.com/ox6VeR8.png)

If this is well received then I plan to remove the old profiler and replace it with a dedicated frametime and FPS display on the status bar on a later PR.

(**Note:** Initially the profiler starts out not displaying anything. To see stuff, enable the groups via the `Groups` menu and then some timers in the `Timer` menu.)

MoochMcGee commented:
That looks a lot like what Xenia uses.

neobrain commented on src/citra_qt/debugger/profiler.cpp:
```diff
@@ -136,3 +143,148 @@ void ProfilerWidget::setProfilingInfoUpdateEnabled(bool enable)
+};
+
+MicroProfileWidget::MicroProfileWidget(QWidget* parent) : QDockWidget(parent) {
+    if (objectName().isEmpty())
```
why the extra check?

yuriks replied:
The other widgets use the same pattern, I just cargo culted it.

neobrain replied:
They do? Like which ones?

yuriks replied:
```
Find all "objectName", Match case, Whole word, Subfolders, Find Results 1, Entire Solution, ""
  C:\Users\yuriks\projects\3ds\citra\src\citra_qt\debugger\profiler.cpp(164):    if (objectName().isEmpty())
  C:\Users\yuriks\projects\3ds\citra\build\src\citra_qt\ui_callstack.h(33):        if (CallStack->objectName().isEmpty())
  C:\Users\yuriks\projects\3ds\citra\build\src\citra_qt\ui_controller_config.h(63):        if (ControllerConfig->objectName().isEmpty())
  C:\Users\yuriks\projects\3ds\citra\build\src\citra_qt\ui_disassembler.h(41):        if (DockWidget->objectName().isEmpty())
  C:\Users\yuriks\projects\3ds\citra\build\src\citra_qt\ui_hotkeys.h(33):        if (hotkeys->objectName().isEmpty())
  C:\Users\yuriks\projects\3ds\citra\build\src\citra_qt\ui_main.h(54):        if (MainWindow->objectName().isEmpty())
  C:\Users\yuriks\projects\3ds\citra\build\src\citra_qt\ui_profiler.h(33):        if (Profiler->objectName().isEmpty())
  C:\Users\yuriks\projects\3ds\citra\build\src\citra_qt\ui_registers.h(33):        if (ARMRegisters->objectName().isEmpty())
  Matching lines: 8    Matching files: 8    Total files searched: 438
```

yuriks replied:
I suppose it's a uic thing then.

neobrain replied:
These are autogenerated files. None of our user-written files use a check there.

neobrain commented on src/citra_qt/debugger/profiler.cpp:
```diff
@@ -136,3 +143,148 @@ void ProfilerWidget::setProfilingInfoUpdateEnabled(bool enable)
+        setObjectName(QStringLiteral("MicroProfile"));
+
+    setWindowTitle(tr("MicroProfile"));
+    resize(1000, 600);
```
I guess there's no way around hardcoding *some* size here, but any particular reason for making the widget this big? A smaller one should be perfectly fine as long as the widget size is saved and restored properly in subsequent runs.

yuriks replied:
I just eyeballed this so that it's large enough to show the whole MP interface without anything overlapping, and small so that it still fits in a 1280x720 screen.

neobrain commented on src/citra_qt/debugger/profiler.h:
```diff
@@ -48,3 +48,17 @@ private slots:
+    void enableTimer(bool enable);
+
+private:
+    QWidget* canvas;
```
This is never used outside of the constructor, hence it should be made a local pointer instead.

neobrain commented on src/citra_qt/debugger/profiler.cpp:
```diff
@@ -2,9 +2,16 @@
 
 #include "common/profiler_reporting.h"
+#include "common/microprofile/microprofile.h"
+#define MICROPROFILEUI_IMPL 1
```
?

neobrain commented on src/citra_qt/debugger/profiler.cpp:
```diff
@@ -2,9 +2,16 @@
 // Licensed under GPLv2 or any later version
 // Refer to the license.txt file included.
 
+#include <QPainter>
```
include order

neobrain commented on src/citra_qt/debugger/profiler.cpp:
```diff
@@ -136,3 +143,148 @@ void ProfilerWidget::setProfilingInfoUpdateEnabled(bool enable)
+    painter.setBackground(Qt::black);
+    painter.eraseRect(rect());
+
+    QFont font("Consolas");
```
Eh, please stick to the font code that we already have instead of adding more ways of retrieving some inconsistent monospace font.

neobrain commented on src/citra_qt/debugger/profiler.cpp:
```diff
@@ -136,3 +143,148 @@ void ProfilerWidget::setProfilingInfoUpdateEnabled(bool enable)
+
+void MicroProfileWidgetCanvas::keyPressEvent(QKeyEvent* ev) {
+    if (ev->key() == Qt::Key_Control) {
+        MicroProfileModKey(1);
```
?

neobrain commented on src/citra_qt/debugger/profiler.cpp:
```diff
@@ -136,3 +143,148 @@ void ProfilerWidget::setProfilingInfoUpdateEnabled(bool enable)
+void MicroProfileDrawText(int nX, int nY, uint32_t nColor, const char* pText, uint32_t nNumCharacters) {
+    mp_painter->setPen(QColor::fromRgb(nColor));
+    // It's impossible to draw a string using a monospaced font with a fixed width per cell in a
+    // way that's reliable across different platforms and fonts as far as I can tell, so draw each
```
I (yuriks)

neobrain commented on src/video_core/command_processor.cpp:
```diff
@@ -43,6 +44,8 @@ static const u32 expand_bits_to_bytes[] = {
     0xffff0000, 0xffff00ff, 0xffffff00, 0xffffffff
 };
 
+MICROPROFILE_DEFINE(GPU_Drawing, "GPU", "Drawing", 0x3333EE);
```
What's the deal with these magic values?

yuriks replied:
They're the color for this profile subcategory in the UI.

neobrain replied:
I guess this could be made more clear by wrapping them in some MICROPROFILE_RGB() macro.

neobrain commented:
I'd rather have microprofile included as a submodule instead of having some (eventually outdated) fork hidden next to our source files.

neobrain commented on src/citra_qt/debugger/profiler.cpp:
```diff
@@ -136,3 +143,148 @@ void ProfilerWidget::setProfilingInfoUpdateEnabled(bool enable)
+    mp_painter->fillRect(nX, nY, nX1 - nX, nY1 - nY, brush);
+}
+
+void MicroProfileDrawLine2D(uint32_t nVertices, float* pVertices, uint32_t nColor) {
```
These should maybe have a comment saying that they are implementations used by the microprofile library (MicroProfileDraw() specifically, I guess?), rather than some citra-internal functions.

neobrain commented on src/citra_qt/main.cpp:
```diff
@@ -64,6 +65,11 @@ GMainWindow::GMainWindow() : emu_thread(nullptr)
 
+    microProfileWidget = new MicroProfileWidget(this);
+    addDockWidget(Qt::LeftDockWidgetArea, microProfileWidget);
+    microProfileWidget->setFloating(true);
```
Is there any point in actually making this a dock widget? Is it realistically possible to use it in a docked state?

Also, I'm not sure how QDockWidget behaves on Windows, but on X11 moving focus to the standalone render window makes all floating dock widgets disappear. Since the profiling widget should always be visible *and* there's dubious point in having it docked at all, it should hence be made a standalone window IMO.

yuriks replied:
That's pretty unfortunate behavior. On Windows the dock widgets are persistent.

My original plan was to not make this a QDockWidget, but in doing so I realized that it gives you a lot of infrastructure that we're using and which I'd have to re-implement separately, namely:
 - Geometry save/restore and remembering the shown state.
 - Action to automatically toggle the window state.
 - Allows making the dock widget a child of the main window for lifetime management.

I agree that it's pretty annoying if it behaves like you describe though, so I'll see what I can do.

neobrain commented:
Looking nice! Made a few comments, nothing major though.

yuriks commented:
@MoochMcGee: That's because they (and also reicast) also use MicroProfile.

@neobrain: MicroProfile is in a mercurial repository, and so can't be directly submoduled. (I also had to tabify it and trim whitespace.) I can copy the files to a directory in externals instead though, I suppose.

yuriks commented:
Addressed some stuff.

archshift commented:
@yuriks perhaps a better idea would be to make another repo under the organization (a la ext-boost) and submodule that, so that it's more manageable? Then we could simply mirror the MicroProfile repo itself.

yuriks commented:
@archshift I don't see any value on doing that over just including the files in the main repo, since microprofile is just 3 files.

yuriks commented:
Updated. I've moved the microprofile files to a directory in externals/ (they're untouched except for tab->spaces and stripping trailing whitespace).

I've also transformed the microprofile widget into a proper dialog, along with some cleanups I noticed while at it.

Should be ready for a re-review.

neobrain commented on externals/microprofile/README.md:
```diff
@@ -0,0 +1,7 @@
+
+It can display profile information in the application, or by generating captures via a minimal built in webserver.
+
+For more information see the project webpage at https://bitbucket.org/jonasmeyer/microprofile
```
Can we have some license and authorship info here?

neobrain replied:
Actually, nevermind, the header files all have proper license headers.

neobrain commented on src/citra_qt/util/util.h:
```diff
@@ -0,0 +1,8 @@
+#include <QFont>
+
+/// Returns a QFont object appropriate to use as a monospace font for debugging widgets, etc.
+QFont GetMonospaceFont();
```
newline at EOF

neobrain commented on src/common/microprofile.cpp:
```diff
@@ -0,0 +1,2 @@
+#define MICROPROFILE_IMPL 1
```
Can we have a comment explaining the point of this file? (i.e. echoing what MICROPROFILE_IMPL does)

neobrain commented on src/common/microprofile.h:
```diff
@@ -0,0 +1,16 @@
+#pragma once
```
license headers?

yuriks replied:
Pretty sure these files aren't even copyrightable, but I'll add them for consistency.

neobrain commented on src/common/microprofile.h:
```diff
@@ -0,0 +1,16 @@
+#define MICROPROFILE_WEBSERVER 0
+#define MICROPROFILE_GPU_TIMERS 0
+#define MICROPROFILE_CONTEXT_SWITCH_TRACE 0
+#define MICROPROFILE_PER_THREAD_BUFFER_SIZE (2048<<12)
```
Is that... 12 GB? Can you add a comment explaining how that is a sensible size?

yuriks replied:
Uh, no? It's 8MB, where did you get 2048 GB from???

neobrain replied:
Derp.... and yeah, 8 MB makes more sense.. :D

yuriks commented:
Updated

yuriks commented:
@neobrain has sort of given this an LGTM. Anyone else have anything to add?

Subv commented:
LGTM

bunnei commented:
This is what I was referring to with the stretching:
![](http://i.imgur.com/KQ88MPH.png)

yuriks commented:
@bunnei: That's expected behavior. As you enabled more timers, or as you have more profiling groups, it add more lines or columns and you can pan around by dragging.

bunnei commented on src/citra_qt/debugger/profiler.cpp:
```diff
@@ -136,3 +148,189 @@ void ProfilerWidget::setProfilingInfoUpdateEnabled(bool enable)
+
+// These functions are called by MicroProfileDraw to draw the interface elements on the screen.
+
+void MicroProfileDrawText(int nX, int nY, uint32_t nColor, const char* pText, uint32_t nNumCharacters) {
```
What's with the non-standard naming convention here?

bunnei replied:
And the other functions throughout this module... I won't bother pointing them out

yuriks replied:
Prototypes copied from the MicroProfile header. Guess I can change them.

bunnei replied:
And the use of `uin32_t` instead of `u32`

bunnei commented on src/core/hw/gpu.cpp:
```diff
@@ -84,6 +85,9 @@ static Math::Vec4<u8> DecodePixel(Regs::PixelFormat input_format, const u8* src_
 }
 
+MICROPROFILE_DEFINE(GPU_DisplayTransfer, "GPU", "DisplayTransfer", 0x6666FF);
+MICROPROFILE_DEFINE(GPU_CmdlistProcessing, "GPU", "Cmdlist Processing", 0x66FF66);
```
Comment what these constants mean?

bunnei replied:
Or what @neobrain said below...

bunnei commented:
Mostly LGTM other than the above minor comments, and ofc being able to maximize the widget per our conversation on IRC

yuriks commented:
Updated addressing @bunnei and @neobrain's comments.

I think I figured out a way to greatly optimize the rendering of the interface, so I want to give that a shot before considering this done.

yuriks commented:
Ok, travis is now fixed too.

bunnei commented:
Looks good
